src/goto-analyzer/show_callsites.cpp

```diff
@@ -0,0 +1,98 @@
+/*******************************************************************\
+
+Module: Write a function call graph to a CSV file
+
+Author: Norbert Manthey, nmanthey@amazon.de
+
+\*******************************************************************/
+
+#include "show_callsites.h"
+
+#include <goto-programs/compute_called_functions.h>
+#include <goto-programs/goto_model.h>
+#include <util/cprover_prefix.h>
+#include <util/json.h>
+#include <util/prefix.h>
+
+#include <fstream>
+#include <iostream>
+#include <sstream>
+
+/** return true, in case name has a CPROVER prefix, or contains a '$' char */
+static bool is_no_program_symbol(const std::string &name)
+{
+  return has_prefix(name, CPROVER_PREFIX) ||
+         name.find('$') != std::string::npos;
+}
+
+/** show_callsites writes a call graph to CSV
+ *
+ * This function creates a map between functions and consumed functions, and
+ * stores this relationship in a CSV file, to be processed outside of
+ * goto-analyzer lateron.
+ *
+ * Functions, whose address is consumed, are listed in a separate section so
+ * that outside reasoning has visibility as well.
+ **/
+void show_callsites(
+  const goto_modelt &goto_model,
+  const std::string &output_filename)
+{
+  // generate all functions which we cannot be sure where they are called
+  // because their address is consumed, and stored into function pointers
+  std::unordered_set<irep_idt> excluded_functions;
+  compute_address_taken_functions(
+    goto_model.goto_functions, excluded_functions);
+
+  std::ofstream output_file;
+
+  if(output_filename != "-")
+  {
+    output_file.open(output_filename);
+    if(!output_file.is_open())
+      std::runtime_error("failed to open callsite output file");
+  }
+
+  std::ostream *output_stream =
+    output_filename == "-" ? &std::cout : &output_file;
```

Comment:
If you can't use the message handler, then store this variable as a reference rather than a pointer.

Reply:
When using a reference, the different types cannot be assigned by the ternary operator, as inheritance requires pointers.

Reply:
References are pointers. Inheritance works with references. The only difference is that reference is never null. Just try it, you'll see :-)